src/backend/distributed/operations/stage_protocol.c

```diff
@@ -825,36 +849,172 @@ UpdateShardStatistics(int64 shardId)
+			continue;
+		}
+
+		for (int64 rowIndex = 0; rowIndex < rowCount; rowIndex++)
+		{
+			bool foundInSet = false;
+			uint64 shardId = ParseIntField(result, rowIndex, 0);
+
+			/* check for the dummy entries we put so that UNION ALL wouldn't complain */
+			if (shardId == INVALID_SHARD_ID)
+			{
+				continue;
+			}
+
+			hash_search(alreadyVisitedShardPlacements,
+						&shardId, HASH_ENTER,
+						&foundInSet);
+			if (foundInSet)
+			{
+				/* We have already updated this placement list */
+				continue;
+			}
+
+			char *minValueResult = PQgetvalue(result, rowIndex, 1);
+			char *maxValueResult = PQgetvalue(result, rowIndex, 2);
+			text *shardMinValue = cstring_to_text(minValueResult);
+			text *shardMaxValue = cstring_to_text(maxValueResult);
+			uint64 shardSize = ParseIntField(result, rowIndex, 3);
+
+			ShardInterval *shardInterval = LoadShardInterval(shardId);
+			Oid relationId = shardInterval->relationId;
+			List *shardPlacementList = ActiveShardPlacementList(shardId);
+
+			UpdateShardSizeAndMinMax(shardId, shardInterval, relationId,
+									 shardPlacementList, shardSize, shardMinValue,
+									 shardMaxValue);
+		}
+		PQclear(result);
+		ForgetResults(connection);
+	}
+	hash_destroy(alreadyVisitedShardPlacements);
+}
+
+
+/*
+ * UpdateShardSizeAndMinMax updates the shardlength (shard size) of the given
+ * shard and its placements in pg_dist_placement, and updates the shard min value
+ * and shard max value of the given shard in pg_dist_shard if the relationId belongs
+ * to an append-distributed table
+ */
+static void
+UpdateShardSizeAndMinMax(uint64 shardId, ShardInterval *shardInterval, Oid relationId,
+						 List *shardPlacementList, uint64 shardSize, text *shardMinValue,
+						 text *shardMaxValue)
+{
+	char storageType = shardInterval->storageType;
+
+	ShardPlacement *placement = NULL;
+
 	/* make sure we don't process cancel signals */
 	HOLD_INTERRUPTS();
 
-	/* update metadata for each shard placement we appended to */
+	/* update metadata for each shard placement */
 	foreach_ptr(placement, shardPlacementList)
 	{
 		uint64 placementId = placement->placementId;
 		int32 groupId = placement->groupId;
 
 		DeleteShardPlacementRow(placementId);
-		InsertShardPlacementRow(shardId, placementId, SHARD_STATE_ACTIVE, shardSize,
+		InsertShardPlacementRow(shardId, placementId, SHARD_STATE_ACTIVE,
+								shardSize,
 								groupId);
 	}
 
 	/* only update shard min/max values for append-partitioned tables */
 	if (IsCitusTableType(relationId, APPEND_DISTRIBUTED))
 	{
 		DeleteShardRow(shardId);
-		InsertShardRow(relationId, shardId, storageType, minValue, maxValue);
+		InsertShardRow(relationId, shardId, storageType, shardMinValue,
+					   shardMaxValue);
 	}
 
 	if (QueryCancelPending)
 	{
-		ereport(WARNING, (errmsg("cancel requests are ignored during metadata update")));
+		ereport(WARNING, (errmsg(
+							  "cancel requests are ignored during metadata update")));
 		QueryCancelPending = false;
 	}
```

Comment:
I think we should remove this if (QueryCancelPending) block  and the HOLD_INTERRUPTS/RESUME_INTERRUPTS. It's extremely outdated code and it might prevent the deadlock detector from actually working.